File: artifacts/api-server/src/routes/risk.ts
import { Router, type IRouter } from "express";
import { db, riskScoresTable, newsSignalsTable, supplierRisksTable } from "@workspace/db";
import { desc } from "drizzle-orm";

const router: IRouter = Router();

router.get("/risk/scores", async (_req, res): Promise<void> => {
  const scores = await db.select().from(riskScoresTable).orderBy(desc(riskScoresTable.score));
  res.json(scores.map(s => ({ ...s, lastUpdated: s.lastUpdated.toISOString() })));
});

router.post("/risk/scores/refresh", async (req, res): Promise<void> => {
  const scores = await db.select().from(riskScoresTable);

  const updated = await Promise.all(
    scores.map(async s => {
      const delta = (Math.random() - 0.4) * 8;
      const newScore = Math.max(0, Math.min(100, s.score + delta));
      const newLevel =
        newScore >= 75 ? "CRITICAL" : newScore >= 50 ? "HIGH" : newScore >= 25 ? "MODERATE" : "LOW";
      const newTrend = delta > 1 ? "increasing" : delta < -1 ? "decreasing" : "stable";

      const [updated] = await db
        .update(riskScoresTable)
        .set({ score: newScore, level: newLevel, trend: newTrend, lastUpdated: new Date() })
        .where({ id: s.id } as any)
        .returning();
      return updated;
    })
  );

  res.json(updated.map(s => ({ ...s, lastUpdated: s.lastUpdated.toISOString() })));
});

router.get("/risk/signals", async (_req, res): Promise<void> => {
  const signals = await db.select().from(newsSignalsTable).orderBy(desc(newsSignalsTable.timestamp)).limit(50);
  res.json(signals.map(s => ({ ...s, timestamp: s.timestamp.toISOString() })));
});

router.get("/risk/suppliers", async (_req, res): Promise<void> => {
  const suppliers = await db.select().from(supplierRisksTable).orderBy(desc(supplierRisksTable.riskScore));
  res.json(suppliers);
});

export default router;

File: artifacts/api-server/src/routes/risk.py
import random
from datetime import datetime
from typing import List
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from sqlalchemy import desc
from ..database import get_db
from ..models import RiskScore, NewsSignal, SupplierRisk
from ..schemas import RiskScoreResponse, NewsSignalResponse, SupplierRiskResponse

router = APIRouter()

@router.get("/risk/scores", response_model=List[RiskScoreResponse])
def get_risk_scores(db: Session = Depends(get_db)):
    scores = db.query(RiskScore).order_by(desc(RiskScore.score)).all()
    return scores

@router.post("/risk/scores/refresh", response_model=List[RiskScoreResponse])
def refresh_risk_scores(db: Session = Depends(get_db)):
    scores = db.query(RiskScore).all()
    updated_scores = []
    
    for s in scores:
        delta = (random.random() - 0.4) * 8
        new_score = max(0.0, min(100.0, s.score + delta))
        
        # Calculate level
        if new_score >= 75:
            new_level = "CRITICAL"
        elif new_score >= 50:
            new_level = "HIGH"
        elif new_score >= 25:
            new_level = "MODERATE"
        else:
            new_level = "LOW"
            
        # Calculate trend
        if delta > 1:
            new_trend = "increasing"
        elif delta < -1:
            new_trend = "decreasing"
        else:
            new_trend = "stable"
            
        s.score = new_score
        s.level = new_level
        s.trend = new_trend
        s.last_updated = datetime.now()
        updated_scores.append(s)
        
    db.commit()
    
    # Return scores sorted by score descending to match Express router
    updated_scores.sort(key=lambda x: x.score, reverse=True)
    return updated_scores

@router.get("/risk/signals", response_model=List[NewsSignalResponse])
def get_news_signals(db: Session = Depends(get_db)):
    signals = db.query(NewsSignal).order_by(desc(NewsSignal.timestamp)).limit(50).all()
    return signals

@router.get("/risk/suppliers", response_model=List[SupplierRiskResponse])
def get_supplier_risks(db: Session = Depends(get_db)):
    suppliers = db.query(SupplierRisk).order_by(desc(SupplierRisk.risk_score)).all()
    return suppliers
